src/Saml2/Provider.php

```diff
@@ -274,69 +418,77 @@ public function user()
             return $this->user;
         }
 
+        $this->receive();
+        $this->ensureSuccessfulStatus();
+
         if ($this->hasInvalidState()) {
             throw new InvalidStateException();
         }
 
+        $this->decryptAssertions();
+
         if ($this->hasInvalidSignature()) {
             throw new InvalidSignatureException();
         }
 
-        $bindingFactory = new BindingFactory();
-        $binding = $bindingFactory->getBindingByRequest($this->request);
-
-        $messageContext = new MessageContext();
-        $binding->receive($this->request, $messageContext);
-
-        $assertion = $messageContext->asResponse()->getFirstAssertion();
+        $assertion = $this->messageContext->asResponse()->getFirstAssertion();
         $attributeStatement = $assertion->getFirstAttributeStatement();
 
         $this->user = new User();
         $this->user->setAssertion($assertion);
         $this->user->map(['id' => $assertion->getSubject()->getNameID()->getValue()]);
 
         if ($attributeStatement) {
-            foreach (['name' => ClaimTypes::NAME, 'email' => ClaimTypes::EMAIL_ADDRESS] as $key => $claim) {
```

Comment:
Nice PR!

I was just preparing to submit a PR extending support for the attributes, Google is not sending `NAME` but sends `GIVEN_NAME` and `SURNAME` [by default].
So I made a simple patch that (tries) to build `name` from these attributes...
 
After reading up about `NAME` I realized it's rather misunderstood, Microsoft states:
>The unique name of the user

Found another person saying:
>The NameID is indeed commonly used for the accountname, but can also contain the emailaddress(when the emailaddress format is used) or other unique attributes. you can give the NameID claim specific formats such as "unspecified" , "emailaddress", "persistant" etc.

>I would not recommend to use the NameID for names(firstname, lastname etc.). It's ment as holder for the unique attribute what is used by the relying party to uniquely identify the user.

https://social.msdn.microsoft.com/Forums/vstudio/en-US/1d36671f-fcb7-48b5-b590-82c6e125af66/correct-usage-of-the-name-claim-httpschemasxmlsoaporgws200505identityclaimsname?forum=Geneva

Thoughts?

Reply:
Here is the patch for reference;
```diff
diff --git a/Provider.php b/Provider.php
index afd40a7..d003b86 100644
--- a/Provider.php
+++ b/Provider.php
@@ -304,6 +304,16 @@ public function user()
                     $this->user->map([$key => $attribute->getFirstAttributeValue()]);
                 }
             }
+
+            if (!$this->user->name) {
+                if ($attribute = $attributeStatement->getFirstAttributeByName(ClaimTypes::GIVEN_NAME)) {
+                    $this->user->name = $attribute->getFirstAttributeValue();
+                }
+
+                if ($attribute = $attributeStatement->getFirstAttributeByName(ClaimTypes::SURNAME)) {
+                    $this->user->name = trim($this->user->name . " " . $attribute->getFirstAttributeValue());
+                }
+            }
 
             $this->user->setRaw($attributeStatement->getAllAttributes());
         }
```

Reply:
NameID (read from the assertion [here](https://github.com/SocialiteProviders/Providers/pull/781/files#diff-b28848b818f26254c5f0d5090662fff2f8406fe28bda138614247d01bab91a18R439)) and `ClaimTypes::NAME` (`http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name`) are not necessarily the same.

The first one always has to be present and this provider requires a persistent NameID from the IdP to make sure it can be used as a unique identifier.

On the other hand the claim `http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name` is just one of the many claims that can be sent and is not necessarily present in the data at all.

Some IdPs can send the same data for the two, that is right, but they can also send different values.

I think we have to keep `ClaimTypes::NAME` as the first option processed by default as the original provider code uses it as well and we would break backwards compatibility if we changed it.

However, with the new configuration options, you can simply remove it in your `config/services.php` like this:
```php
'saml2' => [
    'attribute_map' => [
        'name' => [
            OasisAttributeNameUris::DISPLAY_NAME,
            ClaimTypes::COMMON_NAME,
            OasisAttributeNameUris::COMMON_NAME,
        ],
    ],
],
```